Temp branch for flow control

sparkhi opened the pull request:
Changes in tests, added logging

MancunianSam commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -100,13 +132,20 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+      logger.info(Map("executionName" -> s"""${potentialInput.map(_.executionName)}""", "sourceSystem" -> s"""$systemName""")(s"Writing item to flow control queue table"))
       dependencies.dynamoClient
         .writeItem(
           DADynamoDbWriteItemRequest(
```
This wasn't part of the original plan but can you add `executionName` to the Map that gets passed to `writeItem` so if anything goes wrong we know which step function wrote to the table.

MancunianSam commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -122,16 +161,30 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
     }
 
     for {
+      _ <- IO(logger.info(Map("executionName" -> s"""${potentialInput.map(_.executionName)}""")(s"Starting flow control lambda with input: $potentialInput")))
```
You've got your brackets a bit mixed up here as well. You're doing
```scala
logger.info(Map()("Something"))
```
Which means you're trying to get the value from the Map with a key of "Something" which isn't there so it throws an exception.

You need to change it to 
```scala
logger.info(Map())("Something")
```

All the logger calls are affected I think

techncl commented on scala/lambdas/ingest-flow-control/src/test/scala/uk/gov/nationalarchives/ingestflowcontrol/LambdaTest.scala:
```diff
@@ -115,9 +115,8 @@ class LambdaTest extends AnyFlatSpec with EitherValues:
-    val input = Input("SOM_ExecutionName", "")
 
-    val lambdaRunResult = runLambda(input, initialDynamo, ssmParam, sfnExecutions, predictableRandomNumberSelector())
+    val lambdaRunResult = runLambda(None, initialDynamo, ssmParam, sfnExecutions, predictableRandomNumberSelector())
```
Is it not possible to have a `Some(Input("SOM_ExecutionName", ""))` value?

techncl commented on scala/lambdas/ingest-flow-control/src/test/scala/uk/gov/nationalarchives/ingestflowcontrol/Helpers.scala:
```diff
@@ -89,15 +89,19 @@ object Helpers {
-    override def getItems[T, K](primaryKeys: List[K], tableName: String)(using returnFormat: DynamoFormat[T], keyFormat: DynamoFormat[K]): IO[List[T]] =
-      val firstPK = primaryKeys.head.asInstanceOf[IngestQueuePartitionKey]
-      errors.raise(_.getItem, "Error getting item from dynamo table") >>
+      errors.raise(_.queryItem, "Error getting item from dynamo table") >>
```
1. Maybe "Error querying items in Dynamo table" just so that we no that we queried rather than "read"?
2. Should we add the name of the table to this message?

sparkhi replied:
changed the message. 
No need to add table name. This is a test helper, there is no actual table involved

techncl commented on scala/lambdas/ingest-flow-control/src/test/scala/uk/gov/nationalarchives/ingestflowcontrol/Helpers.scala:
```diff
@@ -89,15 +89,19 @@ object Helpers {
+            values <- Option(requestCondition.attributes.values)
+            map <- values.toMap[String].toOption
+          } yield existing
+            .filter(row => map.get("conditionAttributeValue0").contains(row.sourceSystem))
```
1. Should we use "item" rather than "row"?
2. Is `conditionAttributeValue0` an real value on an item?

sparkhi replied:
I think its a placeholder

techncl commented on scala/lambdas/ingest-flow-control/src/test/scala/uk/gov/nationalarchives/ingestflowcontrol/Helpers.scala:
```diff
@@ -89,15 +89,19 @@ object Helpers {
+          (for {
+            values <- Option(requestCondition.attributes.values)
+            map <- values.toMap[String].toOption
+          } yield existing
```
1. Maybe `existingItems`?
2. What is `existing`? Items that we expect to be in the table? Is so, should it be `expectedItems`?

sparkhi replied:
its a test helper implementation of queryItems. At the beginning of the test, the table starts with whatever the initial state of the table is set in the test. The query items implementation returns what would happen as a result of the action. `Expected`  will be in the context of the test

techncl commented on scala/lambdas/ingest-flow-control/src/test/scala/uk/gov/nationalarchives/ingestflowcontrol/Helpers.scala:
```diff
@@ -109,7 +113,7 @@ object Helpers {
 
-    override def sendTaskSuccess(token: String): IO[Unit] = {
+    override def sendTaskSuccess[T: Encoder](token: String, potentialOutput: Option[T]): IO[Unit] = {
       errors.raise(_.sendTaskSuccess, "Error sending task success to step function") >>
```
1. `potentialOutput` is unused here
2. Should we add the token to the Error message?

sparkhi replied:
As this is just a test implementation, we don't need it but we have to match the signature. similarly, including the token in here won't be of any use. Its just there to test that, if a real `sendTaskSuccess` method fails, we make use of that error. The message here can be anything.

techncl commented on scala/lambdas/ingest-flow-control/src/test/scala/uk/gov/nationalarchives/ingestflowcontrol/Helpers.scala:
```diff
@@ -89,15 +89,19 @@ object Helpers {
-          existing.filter(_.sourceSystem == firstPK.sourceSystem).sortBy(_.queuedAt).map(_.asInstanceOf[T])
+          (for {
+            values <- Option(requestCondition.attributes.values)
+            map <- values.toMap[String].toOption
```
1. Could this not just be `map = values.toMap[String]`
2. The name `map` isn't too clear. When you convert `values` to a `Map`, what is the key and what is value `"S" -> "value"` or `"value" -> "value"` or is it still "attributeName" -> "value"?

sparkhi replied:
tbh, I do not know @MancunianSam what do you reckon?

MancunianSam replied:
`toMap` is a method on `DynamoObject` which, despite it's name, returns `Either[DynamoReadError, Map[String, String]]` which is why this fiddling about with Options is necessary. (We could fiddle about with `Either` but `Option` is fine)

I think considering `requestCondition.attributes.values` is the line directly above, there's enough context for this to make sense so I'd say leave it as is.

techncl reviewed (commented):
Looks good! Just some minor comments, as well as:
1. There is no "queueTablePkFormat read" tests in the `DynamoFormattersTest`
2. You're passing in the executionName to the logger multiple times, could you not use Currying or something?
3. this doesn't have anything to do with this PR but, since `maxConcurrency` is 8, should we set a limit in the FlowControlConfig. we ensure that it is over 0 but not under 9

MancunianSam commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -40,17 +44,30 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
       else if taskStarted then IO.pure(true)
       else
+        logger.info(
+          Map("executionName" -> s"""${potentialInput.map(_.executionName)}""", "sourceSystems" -> s"""${sourceSystems.map(_.systemName).mkString(",")}""")(
```
If it's None then it just prints "None" so that's fine. 
I would go one further here and have a method that logs with the execution name and any other parameters. 

Something like 
```scala
def logWithExecutionName(extraParams: (String, String)*)(msg: String): IO[Unit] =
      logger.info(Map(
        "executionName" -> s"""${potentialInput.map(_.executionName)}""",
      ) ++ extraParams.toMap)(msg)
```

Then you could call
```scala
logWithExecutionName("sourceSystems" -> s"""${sourceSystems.map(_.systemName).mkString(",")}""")(s"Flow control starting on reserved channel")
```

or
```scala
logWithExecutionName()("Iterated over all source systems for probability, none of them have a waiting task, terminating lambda")
```

There are also a couple of instances where you're logging the value of the `IngestQueueTableItem` so maybe a method to do that

```scala
def logItem(item: IngestQueueTableItem)(msg: String): IO[Unit] = {
      logger.info(
        Map(
          "executionName" -> s"""${potentialInput.map(_.executionName)}""",
          "table item" -> s"""${item.sourceSystem}""",
          "queuedAt" -> s"""${item.queuedAt}""",
          "taskToken" -> s"""${item.taskToken}"""
        ))(msg)
    }
```

It's working and I like the Maps in the logger.info call as it adds extra detail but it's quite hard to read with it all inline.

MancunianSam commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -122,16 +161,30 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
       runningExecutions <- dependencies.stepFunctionClient.listStepFunctions(config.stepFunctionArn, Running)
+      _ <- IO(
+        logger.info(
+          Map("executionName" -> s"""${potentialInput.map(_.executionName)}""", "running executions" -> s"""${runningExecutions.map(_.mkString(","))}""")(
```
You can just do `runningExecutions.mkString(",")` here.

sparkhi replied:
yeah, spotted that when making other change, in fact the existing statement was totally wrong as `mkString` was being applied to each element 🤦🏼‍♂️

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -40,17 +44,30 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
       if sourceSystems.isEmpty then IO.pure(taskStarted)
       else if taskStarted then IO.pure(true)
       else
+        logger.info(
```
Changed and also moved the logging calls to help more

sparkhi commented:
> Looks good! Just some minor comments, as well as:
> 
> 1. There is no "queueTablePkFormat read" tests in the `DynamoFormattersTest`
> 2. You're passing in the executionName to the logger multiple times, could you not use Currying or something?
> 3. this doesn't have anything to do with this PR but, since `maxConcurrency` is 8, should we set a limit in the FlowControlConfig. we ensure that it is over 0 but not under 9

1) Done
2) I think its fine 
3) I don't think we should set a limit. It is anyways an arbitrary number. We already ensure that it is > 0

techncl commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -37,28 +41,74 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
-      if sourceSystems.isEmpty then IO.pure(taskStarted)
-      else if taskStarted then IO.pure(true)
+      if sourceSystems.isEmpty then
+        logger.info(Map("taskStarted" -> s"""$taskStarted"""))("list of remaining sourceSystems is empty, nothing more to do on reserved channel") >> IO.pure(taskStarted)
```
You have a bunch of multiline strings throughout the code, but you don't need to do this; you can simply do 
`s"$taskStarted"` or in some cases, just `taskStarted.toString`

techncl commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -172,11 +309,21 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+    for {
+      potentialExecutionName <- c.downField("executionName").as[Option[String]]
+      potentialTaskToken <- c.downField("taskToken").as[Option[String]]
+    } yield {
```
You don't need braces here since it's only one line

sparkhi commented:
> Looks good, quite a few comments (but are mainly minor), as well as these points/questions below
> 
> 1. Would you mind making the "lambda steps" in the README a little more granular because there is quite a lot of logic but there are only 5 steps; Maybe you could break down what each line is doing in the README and in some cases explain why; for example, first we get the execution name from an Option[Input]; why is the input optional?
> 2. How come `taskExecutorName` is always an empty string and why do we need to pass it in
> 3. Line 334 & 335 - You could make these lazy so that they don't run until they are used
> 4. Line 343 & 344 - Do these need to be methods? could they not just be vals since all of your variables are immutable

1) The README is high level and the logic is described in the comment on the methods, I can include it in the ReadMe as well 
2) taskExecutorName is passed as empty string at the start of fresh recursion only.
3)  I'll change it 
4) I'll change it

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -26,39 +31,57 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+        logInfo("Reserved channel: Task success sent by an execution, terminating further computation", taskExecutorName) >>
+          IO.pure(taskExecutorName)
       else
         val currentSystem = sourceSystems.head
```
Done

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -26,39 +31,57 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
         val reservedChannels: Int = flowControlConfig.sourceSystems.find(_.systemName == currentSystem.systemName).map(_.reservedChannels).getOrElse(0)
         val currentExecutionCount = executionsBySystem.getOrElse(currentSystem.systemName, 0)
-        if currentExecutionCount >= reservedChannels then startTaskOnReservedChannel(sourceSystems.tail, executionsBySystem, flowControlConfig, taskStarted)
+        val executionNameForLogging = potentialInput.map(_.executionName).getOrElse("NO_EXECUTION_NAME")
```
Moved

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -26,39 +31,57 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+        val executionNameForLogging = potentialInput.map(_.executionName).getOrElse("NO_EXECUTION_NAME")
+        if currentExecutionCount >= reservedChannels then
+          logInfo(
+            s"Reserved channel: Execution count of $currentExecutionCount exceeds reserved channels",
```
Changed

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -26,39 +31,57 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
-                    IO.pure(true)
-                case None => startTaskOnReservedChannel(sourceSystems.tail, executionsBySystem, flowControlConfig, taskStarted)
-            }
+          logInfo("attempting progress on reserved channel for current system", executionNameForLogging, currentSystem.systemName) >>
```
Changed

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+            val systemToStartTaskOn = sourceSystemEntry._1
+            val systemProbabilitiesStr = sourceSystemProbabilities.map { case (system, range) => s"$system -> (${range.startInclusive}, ${range.endExclusive})" }.mkString(", ")
+            logInfo(
+              s"Selected source system based on probability using lucky dip: $luckyDip in ranges $systemProbabilitiesStr",
```
I don't think we need to, the logging is for monitoring purpose so adding another entry will just add 2 log entries back to back

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
-      dependencies.dynamoClient
-        .writeItem(
-          DADynamoDbWriteItemRequest(
+    def writeItemToQueueTable(flowControlConfig: FlowControlConfig): IO[Unit] = IO.whenA(potentialInput.nonEmpty) {
```
changed

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+      val input = potentialInput.get
+      input.taskToken match
+        case null | "" =>
+          logInfo("Task token is empty, skipping writing item to flow control queue table", input.executionName)
```
I tend to avoid using variables in the log messages so that the code does not fall over when the variable is missing, as the name of the table itself is not  very useful in the logs

techncl commented:
> 2. taskExecutorName is passed as empty string at the start of fresh recursion only.

I could be missing it but I can cannot see a place where you call `startTaskOnReservedChannel` and `taskExecutorName` is not an empty string

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+        case null | "" =>
+          logInfo("Task token is empty, skipping writing item to flow control queue table", input.executionName)
+        case _ =>
+          val sourceSystemName = input.executionName.split("_").head
```
changed

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+          logInfo("Task token is empty, skipping writing item to flow control queue table", input.executionName)
+        case _ =>
+          val sourceSystemName = input.executionName.split("_").head
+          val systemName = flowControlConfig.sourceSystems.find(_.systemName == sourceSystemName).map(_.systemName).getOrElse(default)
```
not quite. All the systems in the config are "supported"

sparkhi replied:
changed

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+        case _ =>
+          val sourceSystemName = input.executionName.split("_").head
+          val systemName = flowControlConfig.sourceSystems.find(_.systemName == sourceSystemName).map(_.systemName).getOrElse(default)
+          val queuedTimeAndExecution = Instant.now.toString + "_" + input.executionName
```
There is a SLIM chance that 2 items could get written at exactly same time, hence we concatenate the execution name to make it unique

techncl replied:
Ah, I just realised it's the sort key

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+          val sourceSystemName = input.executionName.split("_").head
+          val systemName = flowControlConfig.sourceSystems.find(_.systemName == sourceSystemName).map(_.systemName).getOrElse(default)
+          val queuedTimeAndExecution = Instant.now.toString + "_" + input.executionName
+          logInfo("Writing item to ingest queue table", input.executionName, systemName, queuedTimeAndExecution = queuedTimeAndExecution) >>
```
it is extracted from the execution name, so it is there in case we need for debugging purpose

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+              .void
+    }
+
+    def deleteItem(systemName: String, firstItem: IngestQueueTableItem) = {
```
removed

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
-        .void
     }
 
+    def processItems(systemName: String, items: List[IngestQueueTableItem]): IO[String] =
```
changed

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+        case _ =>
+          val item = items.head
+          logInfo(
+            "sending success for the task",
```
changed

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+                  "Task sent successfully, deleting item from table",
+                  item.executionName,
+                  systemName,
+                  queuedTimeAndExecution = s"${item.queuedAtAndExecution}",
```
unwrapped

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -160,9 +293,13 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
-        val rangeEnd = systems.head.probability + rangeStart
-        val newMap = sourceSystemProbabilityRanges ++ Map(systems.head.systemName -> Range(rangeStart, rangeEnd))
-        buildProbabilityRangesMap(systems.tail, rangeEnd, newMap)
+        if systems.head.probability == 0 then {
```
removed

sparkhi commented:
> > 2. taskExecutorName is passed as empty string at the start of fresh recursion only.
> 
> I could be missing it but I can cannot see a place where you call `startTaskOnReservedChannel` and `taskExecutorName` is not an empty string

That may be a bug

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
-      }
-    } yield ()
+      taskSuccessExecutor <-
+        if (runningExecutions.size < flowControlConfig.maxConcurrency) {
```
changed

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+          else
+            val maxRandomValue: Int = sourceSystemProbabilities.values.map(_.endExclusive).max
+            val luckyDip = dependencies.randomInt(1, maxRandomValue)
+            val sourceSystemEntry = sourceSystemProbabilities.find((_, probRange) => probRange.startInclusive <= luckyDip && probRange.endExclusive > luckyDip).get
```
changed

sparkhi commented:
> Just some additional comments, as well as:
> 
> 1. there is a typo in "sendTaskSucess" from the "delete the task from dynamo table when SFN client sendTaskSucess..." test
> 2. Are you testing for this case `!taskExecutorName.isBlank && taskExecutorName != continueProcessingNextSystem`
> 3. There is a lot of branches in the Lambda and I'm worried that we're not testing all of them. I don't know if it too late but do you think that some of the functions like `startTaskOnReservedChannel` and `startTaskBasedOnProbability` would be better off in another file so that it's easier to write unit tests for them/easier to check if we've covered all branches

1)  done
2) I found it was actually a mistake when I introduced the continuation to other systems, that case is never happening because, when a task is progressed, we terminate, now moved the log statement to the correct place and removed this case.
3) Something for future, I'm definitely not inclined to refactor it right now

sparkhi commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+            ) >>
+              dependencies.dynamoClient
+                .queryItems[IngestQueueTableItem](config.flowControlQueueTableName, "sourceSystem" === systemToStartTaskOn)
+                .flatMap { queueTableItems =>
```
changed

MancunianSam commented:
Just one more thing, can you add `ingestFlowControl` here https://github.com/nationalarchives/dr2-ingest/blob/main/scala/lambdas/build.sbt#L19 so it will deploy with everything else please.

techncl commented:
> Just one more thing, can you add `ingestFlowControl` here https://github.com/nationalarchives/dr2-ingest/blob/main/scala/lambdas/build.sbt#L19 so it will deploy with everything else please.

@sparkhi, could you also add an instruction to do this to the README under "Creating a new Lambda.", please?

techncl commented on scala/lambdas/ingest-flow-control/src/main/scala/uk/gov/nationalarchives/ingestflowcontrol/Lambda.scala:
```diff
@@ -74,82 +97,192 @@ class Lambda extends LambdaRunner[Input, Unit, Config, Dependencies] {
+                    "Encountered TaskTimedOutException, deleting item from table",
+                    executionNameForLogging,
+                    systemName,
+                    queuedTimeAndExecution = s"${item.queuedAtAndExecution}"
```
There's one more `item.queuedAtAndExecution` wrapped in braces here
